### ISSUE TITLE
Should mixed class / non-class `=`, `==`, `!=` overloads be supported?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#16830]
In #14180 we made assignment overloads for classes illegal, and in #13048, we have essentially decided that `==` and `!=` should similarly be made illegal for classes.  This issue asks whether these overloads should only be illegal:

* when both operands are classes OR
* when the LHS operand is a class for `=` and when either operand is a class for `==`, `!=` OR
* something else

Put another way, should a user be able to define cases like:

* an assignment to a class from an integer or record?
* an assignment from a class to an integer or record?
* a comparison between a class and an integer or record?

**Comment by mppf** [comment]
I think that they should be illegal and that people should use a `record` if they want to overload these operators for their type.

**Comment by bradcray** [comment]
I was feeling the opposite at a gut level:  That if the goal of making `=`, `==`, and `!=` illegal for classes was to prevent people from "breaking" the pointer nature of classes and it was convenient for a given class to support comparison with integers (where presumably the comparison would be made in terms of the class's fields rather than its identity), maybe we shouldn't strictly outlaw such cases.  I feel less confident about `=` overloads, though I could potentially imagine wanting to assign from a class to a record or maybe even from an integer to a class (where I might require the LHS class to have `const` intent—i.e., the assignment could only change its fields' values, not its identity).

On the other hand, the conservative part of me thinks maybe we should outlaw these cases until someone requests them; or, we could support them with a `--warn-unstable` warning if we didn't have a good reason to outlaw them.

This issue came up when I was working on making `==` illegal for any class/anything combinations and two tests broke that compared classes to scalar values (though admittedly they are not flagship test cases).

**Comment by mppf** [comment]
>  the goal of making =, ==, and != illegal for classes was to prevent people from "breaking" the pointer nature of classes

I agree about that part

I think of `myClassVariable = 1` as misleading, at least, since in C it sets the pointer value to `1`. I don't want to have operations along these lines that invite confusion for people coming from C. IMO the reason for this is the same as why we don't want to allow one to override `myClassVariable = otherClassVariable` - namely, that the operation "looks like it sets a pointer" and that should not be something one can override (since we have records for that).

**Comment by bradcray** [comment]
> since in C it sets the pointer value to 1.

Yes, though modern C compilers will warn against this, and I think conflating our class variables with pointers ignores a lot of things that make them different.

> I don't want to have operations along these lines that invite confusion for people coming from C.

Even so, this could be more of a strong style recommendation (similar to "don't define `+` operators that don't add things together) than a language decision, though.

These responses are just me playing devil's advocate, though, not strong opinions.  I can implement it either way.

I'm noting that C# (which we traditionally used as our reference point for Chapel's class-value interactions) doesn't permit assignment overloads (ever).

**Comment by bradcray** [comment]
I'm also realizing that, taken to its extreme, the "could be confusing to C programmers" potentially applies to other things as well like defining `+` or `<` operators on classes.  I'm much more reluctant to bar those overloads.

**Comment by mppf** [comment]
> I'm also realizing that, taken to its extreme, the "could be confusing to C programmers" potentially applies to other things as well like defining `+` or `<` operators on classes.  I'm much more reluctant to bar those overloads.

I'd bar those too. I guess that makes me extreme :)

**Comment by vasslitvinov** [comment]
I do not see harm in allowing these.

However it makes good sense to disallow them until somebody asks for it.
